static/js/blueslip_stacktrace.ts

```diff
@@ -0,0 +1,122 @@
+});
+
+type CleanStackFrame = {
+    full_path: string;
+    show_path: string;
+    function_name: string;
+    function_scope: string;
+    line_number: number;
+};
+
+export function clean_stackframe_data(data: StackFrame): CleanStackFrame {
+    // If the file is local, just show the filename.
+    // Otherwise, show the full path starting from node_modules.
+    const idx = data.fileName.indexOf('node_modules');
+    let full_path;
+    let show_path;
+
+    if (idx !== -1) {
+        full_path = show_path = data.fileName.slice(idx);
+    } else {
+        full_path = data.fileName.split('/').slice(-3).join('/');
+        show_path = full_path.slice(full_path.lastIndexOf('/') + 1);
+    }
+
+    // Don't show the full scope in function name
+    let full_fn_name;
+    try {
+        full_fn_name = /(?:[a-z_]+\.[jt]s(?:\?\w+)?\.)(.+)/.exec(data.functionName)[1];
+    } catch {
+        full_fn_name = data.functionName;
+    }
+    const fn_name = full_fn_name.split('.');
+
+    return {
+        full_path: full_path,
+        show_path: show_path,
+        line_number: data.lineNumber,
+        function_name: fn_name.pop(),
+        function_scope: fn_name.join('.'),
+    };
+}
+
+// The stacktrace-gps library relies on the mozilla/source-map WASM library,
+// we assume the developer's browser is recent enough to support it.
+const stack_trace_gps = new StackTraceGPS();
+
+export async function display_stacktrace(error: string, stack: string): Promise<void> {
+    const ex = new Error();
+    ex.stack = stack;
+
+    const stackframes: CleanStackFrame[] = await Promise.all(
+        ErrorStackParser.parse(ex).map(async (stackframe: StackFrame) => {
+            const mapped_stackframe = await stack_trace_gps.getMappedLocation(stackframe);
+            return clean_stackframe_data(mapped_stackframe);
+        })
+    );
+
+    const rendered = render_blueslip_stacktrace({ stackframes });
+
+    $('.stacktrace .stacktrace-content').html(rendered);
+    $('.stacktrace .message span').text(error);
+    $('.stacktrace').addClass('show');
+}
+
+function expand_code_context(target: JQuery): void {
+    if (target.css('maxHeight') !== '0px') {
+        target.css('maxHeight', '0px');
+    } else {
+        target.css('maxHeight', target[0].scrollHeight + 'px');
+    }
+    target.toggleClass('show');
+}
+
+export function display_code_context(
+    full_path: string,
+    line_number: number,
+    context_wrapper: JQuery
+): void {
+    if (context_wrapper.children().first().is(':empty')) {
+        $.get('/webpack/app.js.map', async (map: string) => {
+            const sourceContent = await SourceMapConsumer.with(
+                map, null, async (consumer: IndexedSourceMapConsumer) => {
+                    // We don't know the exact file path due to the node_modules
+                    // hash so we have to find it.
+                    for (const source of consumer.sources) {
+                        if (source.match(full_path + '$')) {
```

Comment:
Smashing a dollar sign onto a random string doesn’t make it a regular expression.